### PR TITLE
Drop controller from PUT agent side

### PR DESCRIPTION
**Pull request by riya-17**
Drop controller from PUT agent side

The server no longer relies on the controller
header to PUT, the agent should stop generating
and sending it.

Will be merged after #3193  #3194

### CODE REVIEW

File: lib/pbench/test/unit/agent/task/test_results_push.py [hunk 1]
```diff
@@ -58,7 +58,6 @@ def test_missing_arg():
             args=[
                 TestResultsPush.TOKN_SWITCH,
                 TestResultsPush.TOKN_TEXT,
-                TestResultsPush.CTRL_TEXT,
```
Comment (dbutenhof) [inline lib/pbench/test/unit/agent/task/test_results_push.py:61]: You should probably remove the definition of `CTRL_TEXT` at the top to be sure nobody uses it!

**Review by dbutenhof — DISMISSED** [review]
Looks good if the tests are happy. One "cleanup" suggestion ...

**Review by webbnh — DISMISSED** [review]
Looks good (although, I second Dave's suggestion).

File: lib/pbench/test/unit/agent/task/test_copy_result_tb.py [hunk 2]
```diff
@@ -42,9 +34,7 @@ def test_tarball_nonexistent(self, monkeypatch, agent_logger):
         bad_tarball_name = "nonexistent-tarball.tar.xz"
         expected_error_message = f"Tar ball '{bad_tarball_name}' does not exist"
 
-        monkeypatch.setattr(
-            Path, "exists", self.get_path_exists_mock(bad_tarball_name, False)
-        )
+        monkeypatch.setattr(Path, "exists", lambda self: False)
```
Comment (webbnh) [inline lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]: I freely admit that I'm a big fan of this use of lambda expressions, but I'm curious why you made this change.  The previous code, in addition to providing the `False` value needed to drive the CUT behavior also performed the sanity check that it was being called for the expected file, and this change removes that "blade guard".  What problem was this change intended to solve?

**Review by dbutenhof — APPROVED** [review]
I like the simplification of the lambdas. If that solves the problem, yea!

**Review by webbnh — CHANGES_REQUESTED** [review]
I'm wondering why you decided to "simplify" the mocks for `Path.exists()` in `test_copy_result_tb.py`.

File: lib/pbench/test/unit/agent/task/test_copy_result_tb.py (hunk 2 above)
Comment (riya-17) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]:
```
1@staticmethod
2   def get_path_exists_mock(path: str, result: bool) -> Callable:
3        def mock_func(self: Path) -> bool:
4            assert self.name == path
5            return result
6
7        return mock_func```

It was failing because it was taking file name value as `self.name` value. So, when it was assert in l.4 the `self.name` value that it was taking was `test_copy_result_tb.py` instead of `test_tarball.tar.xz`

Comment (webbnh) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]: This is really annoying.

The previous code (which I really think was better...) was working up through build [#5](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3220/5/consoleFull) of this PR.

Something changed between Feb 08 05:37 (Jenkins time) and Feb 08 14:36, but it's not easy to tell what.  The `main` branch was updated, but only Dashboard files were changed, but I don't see how that could cause this effect.  I cannot tell what was changed on the PR branch, because the previous commit apparently doesn't exist anymore, but looking over the other changes, I don't see anything there which would cause this, either.

I pulled the branch for this PR, and I can reproduce the problem locally.  However, since it doesn't reproduce on the `main` branch, I would really like to understand what is happening here before we carry this forward.

Comment (riya-17) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]: Sure

Comment (webbnh) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]: I've purchased a clue.  (Actually, it was handed to me, but it took me awhile to see if for what it was!)

Looking at the failure in build [#6](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3220/6/console):
```
...
INTERNALERROR> E               File "/var/tmp/jenkins/tox/agent-py36/lib/python3.6/site-packages/_pytest/python.py", line 1726, in _prunetraceback
INTERNALERROR> E                 path, firstlineno = code.path, code.firstlineno
INTERNALERROR> E               File "/var/tmp/jenkins/tox/agent-py36/lib/python3.6/site-packages/_pytest/_code/code.py", line 95, in path
INTERNALERROR> E                 if not p.exists():
INTERNALERROR> E               File "/var/tmp/jenkins/tox/agent-py36/lib/python3.6/site-packages/pbench/test/unit/agent/task/test_copy_result_tb.py", line 27, in mock_func
INTERNALERROR> E                 assert self.name == path
```

Note the `INTERNALERROR` prefix and the fact that the caller is `_pytest`....  What is happening is that some piece of PyTest itself is calling `Path.exists()` and is hitting our mock!!

So, the assertion that I'm so in love with here is pointing out that there is a problem!  It's not obvious to me how removing that assertion (and always returning, e.g., `False`) solves anything.

We need to figure out how/why our mock is being called unexpectedly and address that.

**Review by portante — CHANGES_REQUESTED** [review]
We are dropping controller, but is it in the metadata?

Also, we need to address @webbnh's issue with the unit tests.

File: lib/pbench/test/unit/agent/task/test_copy_result_tb.py (hunk 2 above)
Comment (webbnh) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]: I've narrowed the trigger down to the last assertion in `test_connection_error()`.  The other test scenarios seem to run OK, and that scenario executes past the `with` statement but PyTest loses it during that `assert` statement.

Comment (webbnh) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]: OK, that test is failing, and then PyTest is falling over:
```
::TestCopyResults::test_tarball_nonexistent PASSED
::TestCopyResults::test_with_access[public] PASSED
::TestCopyResults::test_with_access[private] PASSED
::TestCopyResults::test_with_access[None] PASSED
::TestCopyResults::test_connection_error 
tce: about to add
tce: about to monkeypatch
tce: about to with
tcb: in with
tce: about to copy
tce: about to assert
tce: excinfo.value="Cannot connect to 'http://pbench.example.com/api/v1/upload/test_tarball.tar.xz': uh-oh"
tce: expected_error_message="Cannot connect to 'http://pbench.example.com/api/v1/upload/test_tarball.tar.xz'"
tce: trying endswidth(): False
```
Should that be `startswith()` and not `endswith()`?

Comment (webbnh) [reply lib/pbench/test/unit/agent/task/test_copy_result_tb.py:37]: > Should that be `startswith()` and not `endswith()`?

The answer is "yes", and, in fact, you'd already made that correction in the most recent commit, but I'd rolled that one back in order to restore the path-exists mocking.

With the changes to `test_copy_result_tb.py`, except for the `startswith()` change, rolled back, the tests now pass.  So, you should be able to restore the path-exists mocking and move forward.

For now, I propose that we keep the more powerful mock and just live with the ungraceful mode of failure, which we will not encounter when things are working.

**Review by webbnh — CHANGES_REQUESTED** [review]
I don't know that we need to add the controller to the metadata in _this_ PR, but I agree with Peter that that should be done as soon as feasible (but, in another PR, I think).

For this PR, I would like to see the most recent set of changes to `test_copy_result_tb.py` rolled back, except for the necessary correction to the assertion in `test_connection_error()`.

And, we should consider relaxing the constraint on the controller name in results-move.

File: lib/pbench/cli/agent/commands/results/move.py [hunk 3]
```diff
@@ -97,7 +97,6 @@ def execute(self, single_threaded: bool, delete: bool = True) -> int:
 
                 try:
                     crt = CopyResultTb(
-                        self.context.controller,
```
[the remark this answers is not in the document]
Comment (riya-17) [reply lib/pbench/cli/agent/commands/results/move.py:100]: Are we sure we want to remove that? We are providing controller or anyway that is just going to be added in metadata file.

**Comment by riya-17** [comment]
> We are dropping controller, but is it in the metadata?

when we are making tarballs at that time the controller metadata is confirmed in the metadata.log file and if it is not there it is added.

File: lib/pbench/cli/agent/commands/results/move.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/cli/agent/commands/results/move.py:100]: I _think_ all that matters (in this context) is that the Agent can find where the results are.  Apparently, they can be in a non-standard location which the user can point us to using the `controller` option.  If we can find them using that option, then we really don't care (as far as I know) what the value was that the user specified.  (That is, for the moment, there is a requirement that the value be a legal directory name, but that's much less restrictive than requiring it to be a legal host name, which is what the current check does, and we can let the filesystem itself decide whether the value is appropriate -- we don't have to.  Moreover, soon, hopefully, we won't even need it to be a legal directory name.)

So, I don't think we need to validate the value in any way other than to try to use it -- if it works, then everone is happy.  We just need to make sure that we gracefully handle any errors which result from it not working.

File: lib/pbench/agent/results.py [hunk 4]
```diff
@@ -129,28 +129,7 @@ def _unlink_tarball(self, tarball: Path, msg: str) -> str:
...
-          - FileNotFoundError  if the result directory does not have a
-                               metadata.log file
-          - BadMDLogFormat     if the metadata.log file has a pbench.name field
-                               value which does not match the result directory
-                               name.
-          - RuntimeError       if any problems are encountered while creating
-                               or verifying the created tar ball
-        """
-        pbench_run_name = self.result_dir.name
+    def verify_metadata(self, pbench_run_name):
```
Comment (webbnh) [inline lib/pbench/agent/results.py:132]: I missed this on my previous reviews, but it would be appropriate for this function signature to have type hints and a docstring.

File: lib/pbench/cli/agent/commands/results/move.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/cli/agent/commands/results/move.py:100]: In terms of the *agent*, the "controller" is the hostname on which the collection utilities are running when the result directory tree and tarballs are formed. But really it's just an identifier... and on the server side it's just a string.

Right now we're only getting here from the agent CLI which primes this with `hostname`; but when we generalize this as a callable API, why would we care whether the caller is using `hostname` or some other (likely much more useful) identifier? (Especially in a cloud environment where hostname is probably pretty much meaningless.)

On the other hand, I think this is a more strategic/architectural decision than just "oh I'm touching this area of code so I might as well delete the check" and probably doesn't belong in this PR.

File: lib/pbench/agent/results.py [hunk 5]
```diff
@@ -199,6 +178,33 @@ def make_result_tb(self, single_threaded: bool = False) -> TarballRecord:
...
+                               value which does not match the result directory
+                               name.
+          - RuntimeError       if any problems are encountered while creating
+                               or verifying the created tar ball
+        """
+        pbench_run_name = self.result_dir.name
+        try:
+            self.verify_metadata(pbench_run_name)
+        except Exception:
+            raise
```
Comment (webbnh) [inline lib/pbench/agent/results.py:206]: Here's another one which I previously missed:  it doesn't look like this `try` block has any effect!  It catches basically all exceptions and immediately reraises them, which is exactly what would happen if it weren't there.

You should probably remove it.

File: lib/pbench/test/unit/agent/task/test_copy_result_tb.py [hunk 6]
```diff
@@ -37,37 +37,15 @@ def mock_func(self: Path, mode: str) -> io.IOBase:
...
-
     def test_tarball_nonexistent(self, monkeypatch, agent_logger):
         """Test error when the tarball file does not exist"""
         bad_tarball_name = "nonexistent-tarball.tar.xz"
         expected_error_message = f"Tar ball '{bad_tarball_name}' does not exist"
 
-        monkeypatch.setattr(
-            Path, "exists", self.get_path_exists_mock(bad_tarball_name, False)
-        )
+        monkeypatch.setattr(Path, "exists", lambda self: False)
```
Comment (webbnh) [inline lib/pbench/test/unit/agent/task/test_copy_result_tb.py:45]: Now that you've restored `get_path_exists_mock()`, is there a reason why we're not using it here once again?

**Review by dbutenhof — APPROVED** [review]
@portante writes:
> We are dropping controller, but is it in the metadata?

What does this mean? First off, it should still be in `metadata.log` since I don't believe Riya has touched any of that code. And since we're only talking about agent-generated tarballs here, there's always a `metadata.log`.

More strategically, though... do we want a way for non-agent client to specify a "controller metadata" in some formalized way? I defined `server.origin` and `server.archiveonly` and removed `controller` from the dataset table. Do we really *want* to have some permanent concept analogous to "controller" going forward, aside from the Pbench Agent `metadata.log`? I thought the decision was that we wouldn't. It's "just metadata", and in most modern environments "hostname where we ran the collection tool" isn't a particularly meaningful concept.

Aside from the peculiar (and relatively transient) Pbench Satellite put-shim use, anyone's welcome to use `server.origin` to record "where this came from", and for put-shim we've got that in addition to our standard `metadata.log`. Do we need more?

**Review by webbnh — COMMENTED** [review]
There's one place where the code using `get_path_exists_mock()` wasn't restored.  And, as long as you're making the change to fix that, I found two other things which I'd previously missed which should probably be fixed, too.

File: lib/pbench/cli/agent/commands/results/move.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/cli/agent/commands/results/move.py:100]: >  I think this is a more strategic/architectural decision than just "oh I'm touching this area of code so I might as well delete the check" and probably doesn't belong in this PR.

I don't care strongly, but I remember being bitten by this check (or one like it) the last time we tried to do bulk uploads of existing results...and I'd rather not be bitten by it again, especially since I think we all agree that the check is overzealous.  And, I can't really say whether it belongs in this PR, but it certainly fits under the _title_ of this PR.  😉

**Review by webbnh — APPROVED** [review]
👍

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/agent/results.py [hunk 7]
```diff
@@ -129,28 +129,8 @@ def _unlink_tarball(self, tarball: Path, msg: str) -> str:
...
-          - FileNotFoundError  if the result directory does not have a
-                               metadata.log file
-          - BadMDLogFormat     if the metadata.log file has a pbench.name field
-                               value which does not match the result directory
-                               name.
-          - RuntimeError       if any problems are encountered while creating
-                               or verifying the created tar ball
-        """
-        pbench_run_name = self.result_dir.name
+    def verify_metadata(self, pbench_run_name: str):
```
Comment (portante) [inline lib/pbench/agent/results.py:132]: What was the reason to make this change to add `verify_metadata`?  There doesn't seem to be another caller besides `make_result_tb()?

Seems like this change has nothing to do with dropping the controller from the PUT on the agent side.

File: lib/pbench/cli/agent/commands/results/move.py (hunk 3 above)
Comment (portante) [inline lib/pbench/cli/agent/commands/results/move.py:100]: Don't we also need to remove the `--controller` option at line 179 below?

**Review by portante — CHANGES_REQUESTED** [review]

File: lib/pbench/cli/agent/commands/results/move.py (hunk 3 above)
Comment (webbnh) [reply lib/pbench/cli/agent/commands/results/move.py:100]: Making the result tarball still needs to know where the result directory is, so that switch still has value (it is used to locate non-standard directory trees).

Comment (riya-17) [reply lib/pbench/cli/agent/commands/results/move.py:100]: that is an optional option. That is basically used for storing the controller metadata and verifying that in metadata.log file.

File: lib/pbench/agent/results.py (hunk 7 above)
Comment (riya-17) [reply lib/pbench/agent/results.py:132]: yes, it doesn't have anything to do with this change but it was good to make because that is a long function and it fits well in different functional entity.

File: lib/pbench/cli/agent/commands/results/move.py (hunk 3 above)
Comment (portante) [reply lib/pbench/cli/agent/commands/results/move.py:100]: Why is this option only available on move and not push?

Comment (webbnh) [reply lib/pbench/cli/agent/commands/results/move.py:100]: Because `-move` creates the tarball while `-push` only uploads it?  My understanding is that we need to know the location (and/or be given the controller to add to the metadata) when we create the tarball, but, once the tarball exists, the controller is "baked in", so `-push` doesn't need (or want) to know about it.  (But...I could be confused....)
